File: Python/lab9/solutions/producer.py
import random
from confluent_kafka import avro
import logging
import uuid
from confluent_kafka.avro import AvroProducer

def load_avro_schema_from_file(schema_file="DisasterValue.avsc"):
    key_schema_string = """
    {"type": "string"}
    """

    key_schema = avro.loads(key_schema_string)
    value_schema = avro.load(schema_file)

    return key_schema, value_schema

def create_message(disaster, intensity_scale, intensity, recommendations):
    return {
        'disasterType': disaster,
        'intensity': {
            'scale': intensity_scale,
            'measurement': intensity
        },
        'recommendations': recommendations
  }

def send_record():
    key_schema, value_schema = load_avro_schema_from_file()

    producer_config = {
        "bootstrap.servers": 'localhost:9092',
        "schema.registry.url": 'http://localhost:8081'
    }

    producer = AvroProducer(producer_config, default_key_schema=key_schema, default_value_schema=value_schema)
    topic = "disasters"
    for _ in range(1000):
        key = str(uuid.uuid4())
        disaster = random.choice(["flood", "hurricane"])
        intensity = random.randint(1, 10)
        recommendations_dict = {
            'flood': ["turn off lights", "evacuate"],  
            'hurricane': ["run"]
        }
        scales_dict = {
            'flood': "meters",
            'hurricane': "richter"
        }
        value = create_message(disaster=disaster, intensity=intensity, recommendations=recommendations_dict[disaster], intensity_scale=scales_dict[disaster])

        try:
            producer.produce(topic=topic, key=key, value=value)
        except Exception as e:
            print(f"Exception while producing record value - {value} to topic - {topic}: {e}")
        else:
            print(f"Successfully producing record value - {value} to topic - {topic}")

    producer.flush()


if __name__ == "__main__":
    send_record()

File: JS/lab9/solutions/producer.js
const { KafkaAvro } = require('kafkajs-avro')

const kafka = new KafkaAvro({
  brokers: ['localhost:9092', 'localhost:9093'],
  clientId: 'lab9-producer',
  avro: {
    url: "http://localhost:8081"
  }
})

const producer = kafka.avro.producer()
const topic = 'disasters'
const typesOfDisasters = ['flood', 'hurricane']

const getIntensity = () => Math.floor(Math.random()* 10)
const recommendationsDict = {
  flood: ["turn off lights", "evacuate"],  
  hurricane: ["run"]
}
const scalesDict = {
  flood: "meters",
  hurricane: "richter"
}
const createMessage = (disaster, intensity, scale, recommendations) => ({
  subject: "test.DisasterValue",
  version: "1",
  value: {
    disasterType: `${disaster}`,
    intensity: {
      scale: scale,
      measurement: intensity
    },
    recommendations: recommendations
  }
})

const sendMessage = () => {
  intensity = getIntensity()
  disaster = typesOfDisasters[Math.floor(Math.random()*typesOfDisasters.length)]
  recommendationToGive = recommendationsDict[disaster]
  scale = scalesDict[disaster]
  return producer
    .send({
      topic,
      messages: Array(1)
        .fill()
        .map(_ => createMessage(disaster, intensity, scale, recommendationToGive)),
    })
    .then(console.log)
    .catch(e => console.error(`[example/producer] ${e.message}`, e))
}

const run = async () => {
  await producer.connect()
  setInterval(sendMessage, 3000)
}

run().catch(e => console.error(`[example/producer] ${e.message}`, e))
